File: src/oca/java/test/VirtualRouterTest.java
/*******************************************************************************
 * Copyright 2002-2015, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
import static org.junit.Assert.assertTrue;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.group.Group;
import org.opennebula.client.template.*;
import org.opennebula.client.vrouter.*;
import org.opennebula.client.vm.*;
import org.opennebula.client.user.User;
import org.opennebula.client.vm.VirtualMachine;



public class VirtualRouterTest
{
    private static VirtualRouter vrouter;
    private static VirtualRouterPool vrouterPool;

    private static Client client;

    private static OneResponse res;
    private static String name = "new_test_vrouter";


    private static String template_str =
        "NAME = \"" + name + "\"\n" +
        "ATT1 = \"VAL1\"\n" +
        "ATT2 = \"VAL2\"";

    /**
     * @throws java.lang.Exception
     */
    @BeforeClass
    public static void setUpBeforeClass() throws Exception
    {
        client          = new Client();
        vrouterPool     = new VirtualRouterPool(client);
    }

    /**
     * @throws java.lang.Exception
     */
    @AfterClass
    public static void tearDownAfterClass() throws Exception
    {
    }

    /**
     * @throws java.lang.Exception
     */
    @Before
    public void setUp() throws Exception
    {
        res = VirtualRouter.allocate(client, template_str);

        int oid = res.isError() ? -1 : Integer.parseInt(res.getMessage());
        vrouter = new VirtualRouter(oid, client);
    }

    /**
     * @throws java.lang.Exception
     */
    @After
    public void tearDown() throws Exception
    {
        vrouter.delete();
    }

    @Test
    public void allocate()
    {
        vrouter.delete();

        res = VirtualRouter.allocate(client, template_str);
        assertTrue( res.getErrorMessage(), !res.isError() );

        int oid = res.isError() ? -1 : Integer.parseInt(res.getMessage());
        vrouter = new VirtualRouter(oid, client);


        vrouterPool.info();

        boolean found = false;
        for(VirtualRouter temp : vrouterPool)
        {
            found = found || temp.getName().equals(name);
        }

        assertTrue( found );
    }

    @Test
    public void info()
    {
        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.getName().equals(name) );
    }

    @Test
    public void publish()
    {
        res = vrouter.publish();
        assertTrue( res.getErrorMessage(), !res.isError() );

        vrouter.info();
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_U").equals("1") );
    }

    @Test
    public void unpublish()
    {
        res = vrouter.unpublish();
        assertTrue( res.getErrorMessage(), !res.isError() );

        vrouter.info();
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_U").equals("0") );
    }

    @Test
    public void chmod()
    {
        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        String owner_a = vrouter.xpath("PERMISSIONS/OWNER_A");
        String group_a = vrouter.xpath("PERMISSIONS/GROUP_A");

        res = vrouter.chmod(0, 1, -1, 1, 0, -1, 1, 1, 0);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_U").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_M").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_A").equals(owner_a) );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_A").equals(group_a) );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_U").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_M").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_A").equals("0") );
    }

    @Test
    public void chmod_octet()
    {
        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vrouter.chmod(640);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_U").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_M").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_A").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_A").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_U").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_M").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_A").equals("0") );

        res = vrouter.chmod("147");
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_U").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_M").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/OWNER_A").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_U").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_M").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/GROUP_A").equals("0") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_U").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_M").equals("1") );
        assertTrue( vrouter.xpath("PERMISSIONS/OTHER_A").equals("1") );
    }

    @Test
    public void attributes()
    {
        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.xpath("NAME").equals(name) );
    }

    @Test
    public void delete()
    {
        res = vrouter.delete();
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vrouter.info();
        assertTrue( res.isError() );
    }

    @Test
    public void chown()
    {
        // Create a new User and Group
        res = User.allocate(client, "template_test_user", "password");
        assertTrue( res.getErrorMessage(), !res.isError() );

        int uid = Integer.parseInt(res.getMessage());

        res = Group.allocate(client, "template_test_group");
        assertTrue( res.getErrorMessage(), !res.isError() );

        int gid = Integer.parseInt(res.getMessage());

        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.uid() == 0 );
        assertTrue( vrouter.gid() == 0 );

        res = vrouter.chown(uid, gid);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.uid() == uid );
        assertTrue( vrouter.gid() == gid );

        res = vrouter.chgrp(0);

        res = vrouter.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vrouter.uid() == uid );
        assertTrue( vrouter.gid() == 0 );
    }

    @Test
    public void instantiate()
    {
        VirtualMachinePool vmPool = new VirtualMachinePool(client);

        res = vmPool.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmPool.getLength() == 0 );

        String tmpl_str =
            "NAME = vrtemplate\n"+
            "CPU = 0.1\n"+
            "MEMORY = 64\n";

        res = Template.allocate(client, tmpl_str);
        assertTrue( res.getErrorMessage(), !res.isError() );

        int tmplid = Integer.parseInt(res.getMessage());

        res = vrouter.instantiate(3, tmplid);
        assertTrue( res.getErrorMessage(), !res.isError() );

        res = vmPool.info();
        assertTrue( res.getErrorMessage(), !res.isError() );

        assertTrue( vmPool.getLength() == 3 );
    }
}

File: src/oca/go/src/goca/virtualrouter_test.go
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2025, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/*--------------------------------------------------------------------------- */

package goca

import (
	. "gopkg.in/check.v1"

	"github.com/OpenNebula/one/src/oca/go/src/goca/parameters"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/shared"
	vn "github.com/OpenNebula/one/src/oca/go/src/goca/schemas/virtualnetwork"
	vnkeys "github.com/OpenNebula/one/src/oca/go/src/goca/schemas/virtualnetwork/keys"
	vrouter "github.com/OpenNebula/one/src/oca/go/src/goca/schemas/virtualrouter"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/virtualrouter/keys"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vm"
	vmkeys "github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vm/keys"
)

type VRSuite struct {
	ID       int
}

var _ = Suite(&VRSuite{})

func (s *VRSuite) SetUpSuite(c *C) {
	var vr_name string = "new_vr"

	vr_template := vrouter.NewTemplate()
	vr_template.Add(keys.Name, vr_name)
	vr_template.AddPair("ATT1", "VAL1")
	vr_template.AddPair("ATT2", "VAL2")

	//Create VirtualRouter
	vr_id, err := testCtrl.VirtualRouters().Create(vr_template.String())
	c.Assert(err, IsNil)

	s.ID = vr_id
}

func (s *VRSuite) TearDownSuite(c *C) {
	// Delete Virtul Network
	testCtrl.VirtualRouter(s.ID).Delete()
}

func (s *VRSuite) TestGetByNameAndID(c *C) {
	// Get Virtual Network by ID
	vrC := testCtrl.VirtualRouter(s.ID)
	vr, err := vrC.Info(false)
	c.Assert(err, IsNil)
	c.Assert(vr.ID, Equals, s.ID)

	// Get Virtual Network by Name
	id, err := testCtrl.VirtualRouterByName(vr.Name)
	c.Assert(err, IsNil)
	c.Assert(vr.ID, Equals, id)
}

func (s *VRSuite) TestUpdate(c *C) {
	// Update
	vrC := testCtrl.VirtualRouter(s.ID)
	err := vrC.Update("ATT3=VAL3", parameters.Merge)
	c.Assert(err, IsNil)

	vr, err := vrC.Info(false)
	c.Assert(err, IsNil)

	actual_1, err := vr.Template.GetStr("ATT1")
	c.Assert(err, IsNil)
	c.Assert(actual_1, Equals, "VAL1")

	actual_3, err := vr.Template.GetStr("ATT3")
	c.Assert(err, IsNil)
	c.Assert(actual_3, Equals, "VAL3")
}

func (s *VRSuite) TestRename(c *C) {
	vrC := testCtrl.VirtualRouter(s.ID)
	vrC.Rename("new_name")

	vr, err := vrC.Info(false)
	c.Assert(err, IsNil)
	c.Assert(vr.Name, Equals, "new_name");
}

func (s *VRSuite) TestChmod(c *C) {
	new_permissions := shared.Permissions{1, 1, 1, 1, 1, 1, 1, 1, 1}

	vrC := testCtrl.VirtualRouter(s.ID)

	err := vrC.Chmod(new_permissions)

	c.Assert(err, IsNil)

	vr, err := vrC.Info(false)

	c.Assert(err, IsNil)
	c.Assert(*vr.Permissions, Equals, new_permissions);
}

func (s *VRSuite) TestChown(c *C) {
	// Test only if the call exists, no real change
	vrC := testCtrl.VirtualRouter(s.ID)
	err := vrC.Chown(1, 1)

	c.Assert(err, IsNil)

	vr, err := vrC.Info(false)
	c.Assert(err, IsNil)
	c.Assert(vr.UID, Equals, 1);
	c.Assert(vr.GID, Equals, 1);
}

func (s *VRSuite) TestLock(c *C) {
	// Lock
	vrC := testCtrl.VirtualRouter(s.ID)
	err := vrC.Lock(shared.LockUse)
	c.Assert(err, IsNil)

	vr, err := vrC.Info(false)
	c.Assert(err, IsNil)
	c.Assert(vr.LockInfos.Locked, Equals, 1);

	// Unlock
	err = vrC.Unlock()
	c.Assert(err, IsNil)

	vr, err = vrC.Info(false)
	c.Assert(err, IsNil)
	c.Assert(vr.LockInfos, IsNil)
}

func (s *VRSuite) TestInstantiate(c *C) {
	vrTmpl := vm.NewTemplate()
	vrTmpl.Add(vmkeys.Name, "vrtemplate")
	vrTmpl.Add(vmkeys.VRouter, "YES")
	vrTmpl.CPU(0.1).Memory(64)

	tmpl_id, err := testCtrl.Templates().Create(vrTmpl.String())
	c.Assert(err, IsNil)

	//Instantiate VirtualRouter
	vrC := testCtrl.VirtualRouter(s.ID)
	vrC.Instantiate(1, int(tmpl_id), "vr_test_go", false, "")

	id, err := testCtrl.VMs().ByName("vr_test_go")
	c.Assert(err, IsNil)

	err = testCtrl.VM(id).TerminateHard()
	c.Assert(err, IsNil)

	err = testCtrl.Template(tmpl_id).Delete()
	c.Assert(err, IsNil)
}

func (s *VRSuite) TestNIC(c *C) {
	vn_tmpl := vn.NewTemplate()
	vn_tmpl.Add(vnkeys.Name, "go-net")
	vn_tmpl.Add(vnkeys.Bridge, "vbr0")
	vn_tmpl.Add(vnkeys.VNMad, "dummy")

	vnet_id, _ := testCtrl.VirtualNetworks().Create(vn_tmpl.String(), 0)
	vnetC := testCtrl.VirtualNetwork(vnet_id)
	WaitState(c, vnetC, "READY")

	nic_tmpl := shared.NewNIC()
	nic_tmpl.Add(shared.Network, "go-net")

	//Attach nic to VirtualRouter
	vrC := testCtrl.VirtualRouter(s.ID)
	err := vrC.AttachNic(nic_tmpl.String())
	c.Assert(err, IsNil)

	vr, err := vrC.Info(false)
	c.Assert(err, IsNil)

	nics := vr.Template.GetVectors(string(shared.NICVec))
	c.Assert(len(nics), Equals, 1)

	actualNetName, _ := nics[0].GetStr("NETWORK")
	c.Assert(actualNetName, Equals, "go-net")

	err = vnetC.Delete()
	c.Assert(err, IsNil)

	//Detach nic from VirtualRouter
	err = vrC.DetachNic(0)
	c.Assert(err, IsNil)
}
